<?xml version="1.0"?>
<launch>
  <!-- Arguments -->
  <arg name="model" default="$(env TURTLEBOT3_MODEL)" doc="model type [burger, waffle, waffle_pi]"/>
  <arg name="map_file" default="$(find robo_navi)/maps/maze.yaml"/>
  <arg name="open_rviz" default="false"/>
  <arg name="move_forward_only" default="false"/>
  <arg name="robot_name"/>
  <arg name="x"/>
  <arg name="y"/>
  <arg name="a"/>

  <group ns = "$(arg robot_name)">
    <!-- Map server -->
    <node pkg="map_server" name="map_server" type="map_server" args="$(arg map_file)"/>

    <!-- AMCL -->
    <node pkg="amcl" type="amcl" name="amcl">

      <param name="min_particles" value="500"/>
      <param name="max_particles" value="3000"/>
      <param name="kld_err" value="0.02"/>
      <param name="update_min_d" value="0.20"/>
      <param name="update_min_a" value="0.20"/>
      <param name="resample_interval" value="1"/>
      <param name="transform_tolerance" value="0.5"/>
      <param name="recovery_alpha_slow" value="0.00"/>
      <param name="recovery_alpha_fast" value="0.00"/>
      <param name="initial_pose_x" value="$(arg x)"/>
      <param name="initial_pose_y" value="$(arg y)"/>
      <param name="initial_pose_a" value="$(arg a)"/>
      <param name="gui_publish_rate" value="50.0"/>

      <remap from="scan" to="scan"/>
      <param name="laser_max_range" value="3.5"/>
      <param name="laser_max_beams" value="180"/>
      <param name="laser_z_hit" value="0.5"/>
      <param name="laser_z_short" value="0.05"/>
      <param name="laser_z_max" value="0.05"/>
      <param name="laser_z_rand" value="0.5"/>
      <param name="laser_sigma_hit" value="0.2"/>
      <param name="laser_lambda_short" value="0.1"/>
      <param name="laser_likelihood_max_dist" value="2.0"/>
      <param name="laser_model_type" value="likelihood_field"/>

      <param name="odom_model_type" value="diff"/>
      <param name="odom_alpha1" value="0.1"/>
      <param name="odom_alpha2" value="0.1"/>
      <param name="odom_alpha3" value="0.1"/>
      <param name="odom_alpha4" value="0.1"/>
      <param name="odom_frame_id" value="/$(arg robot_name)/odom"/>
      <param name="base_frame_id" value="/$(arg robot_name)/base_footprint"/>

    </node>



    <!-- move_base -->
    <!-- Arguments -->
    <arg name="cmd_vel_topic" default="cmd_vel" />
    <arg name="odom_topic" default="odom" />

    <!-- move_base -->
    <node pkg="move_base" type="move_base" respawn="false" name="move_base" output="screen">
      <param name="base_local_planner" value="dwa_local_planner/DWAPlannerROS" />
      <rosparam file="$(find robo_navi)/param/local_costmap_params.yaml" command="load" />
      <rosparam file="$(find robo_navi)/param/global_costmap_params.yaml" command="load" />
      <rosparam file="$(find robo_navi)/param/move_base_params.yaml" command="load" />
      <rosparam file="$(find robo_navi)/param/dwa_local_planner_params_$(arg model).yaml" command="load" />
      <rosparam file="$(find robo_navi)/param/costmap_common_params_$(arg model).yaml" command="load" ns="global_costmap" />
      <rosparam file="$(find robo_navi)/param/costmap_common_params_$(arg model).yaml" command="load" ns="local_costmap" />
      <!-- Set tf_prefix for frames explicity, overwriting defaults -->
      <param name="global_costmap/scan/sensor_frame" value="$(arg robot_name)/base_scan"/>
      <param name="global_costmap/obstacle_layer/scan/sensor_frame" value="$(arg robot_name)/base_scan"/>
      <param name="global_costmap/global_frame" value="map"/>
      <param name="global_costmap/robot_base_frame" value="$(arg robot_name)/base_footprint"/>
      <param name="local_costmap/scan/sensor_frame" value="$(arg robot_name)/base_scan"/>
      <param name="local_costmap/obstacle_layer/scan/sensor_frame" value="$(arg robot_name)/base_scan"/>
      <param name="local_costmap/global_frame" value="$(arg robot_name)/odom"/>
      <param name="local_costmap/robot_base_frame" value="$(arg robot_name)/base_footprint"/>
      <!-- <remap from="/tb3_0/cmd_vel" to="$(arg cmd_vel_topic)"/>
    <remap from="/tb3_0/odom" to="$(arg odom_topic)"/> -->
      <param name="DWAPlannerROS/min_vel_x" value="0.0" if="$(arg move_forward_only)" />
      <!-- <remap from="map" to="/map"/> -->
    </node>
  </group>


  <!-- rviz -->
  <group if="$(arg open_rviz)">
    <node pkg="rviz" type="rviz" name="rviz" required="true" args="-d $(find robo_navi)/rviz/robo_navi.rviz"/>
  </group>
</launch>
